src/rulesets/LinearAlgebra/matfun.jl

```diff
@@ -0,0 +1,270 @@
+# matrix functions of dense matrices
+
+# NOTE: for matrix functions whose power series representation has real coefficients,
+# the pullback and pushforward are related by an adjoint.
+# Specifically, if the pushforward of f(A) is (f_*)_A(ΔA), then the pullback at Y=f(A) is
+# (f^*)_Y(ΔY) = (f_*)_{A'}(ΔY) = ((f_*)_A(ΔY'))'
+# So we reuse the code from the pushforward to implement the pullback.
+
+"""
+    _matfun(f, A) -> (Y, intermediates)
+
+Compute the matrix function `Y=f(A)` for matrix `A`.
+The function returns a tuple containing the result and a tuple of intermediates to be
+reused by `_matfun_frechet` to compute the Fréchet derivative.
+Note that any function `f` used with this **must** have a `frule` defined on it.
+"""
+_matfun
+
+"""
+    _matfun!(f, A) -> (Y, intermediates)
+
+Similar to [`_matfun`](@ref), but where `A` may be overwritten.
+"""
+_matfun!
+
+"""
+    _matfun_frechet(f, A, Y, ΔA, intermediates)
+
+Compute the Fréchet derivative of the matrix function `Y=f(A)`, where the Fréchet derivative
+of `A` is `ΔA`, and `intermediates` is the second argument returned by `_matfun`.
+"""
+_matfun_frechet
+
+"""
+    _matfun_frechet!(f, A, Y, ΔA, intermediates)
+
+Similar to [`_matfun_frechet`](@ref), but where `ΔA` may be overwritten.
+"""
+_matfun_frechet!
+
+#####
+##### `exp`/`exp!`
+#####
+
+function frule((_, ΔA), ::typeof(LinearAlgebra.exp!), A::StridedMatrix{<:BlasFloat})
+    if ishermitian(A)
+        hermX, ∂hermX = frule((Zero(), ΔA), exp, Hermitian(A))
+        X = LinearAlgebra.copytri!(parent(hermX), 'U', true)
+        if ∂hermX isa LinearAlgebra.RealHermSymComplexHerm
+            ∂X = LinearAlgebra.copytri!(parent(∂hermX), 'U', true)
+        else
+            ∂X = ∂hermX
+        end
+    else
+        X, intermediates = _matfun!(exp, A)
+        ∂X = _matfun_frechet!(exp, A, X, ΔA, intermediates)
+    end
+    return X, ∂X
+end
+
+function rrule(::typeof(exp), A0::StridedMatrix{<:BlasFloat})
+    # TODO: try to make this more type-stable
+    if ishermitian(A0)
+        # call _matfun instead of the rrule to avoid hermitrizing ∂A in the pullback
+        hermA = Hermitian(A0)
+        hermX, hermX_intermediates = _matfun(exp, hermA)
+        function exp_pullback_hermitian(ΔX)
+            ∂hermA = _matfun_frechet(exp, hermA, hermX, ΔX, hermX_intermediates)
+            ∂hermA isa LinearAlgebra.RealHermSymComplexHerm || return NO_FIELDS, ∂hermA
+            return NO_FIELDS, parent(∂hermA)
+        end
+        return LinearAlgebra.copytri!(parent(hermX), 'U', true), exp_pullback_hermitian
+    else
+        A = copy(A0)
+        X, intermediates = _matfun!(exp, A)
+        function exp_pullback(ΔX)
+            ΔX′ = copy(adjoint(ΔX))
```

Comment:
Perhaps. It might make sense to explicitly introduce a `_matfun_frechet_adjoint` and in-place version that handle the adjoint logic, which will be a little different for every matrix type.